File: src/lib/structure-map-executor.ts
import { StructureMap, ExecutionResult, ExecutionOptions, EnhancedExecutionResult } from '../types';
import { ValidationService } from './validation-service';
import { ConceptMapService } from './conceptmap-service';
import { ValueSetService } from './valueset-service';
import { CodeSystemService } from './codesystem-service';

/**
 * StructureMap execution engine - executes StructureMaps on input data
 */
export class StructureMapExecutor {
  private validationService: ValidationService;
  private conceptMapService?: ConceptMapService;
  private valueSetService?: ValueSetService;
  private codeSystemService?: CodeSystemService;

  constructor() {
    this.validationService = new ValidationService();
  }

  /**
   * Set terminology services for advanced transformation support
   */
  setTerminologyServices(
    conceptMapService: ConceptMapService,
    valueSetService: ValueSetService,
    codeSystemService: CodeSystemService
  ): void {
    this.conceptMapService = conceptMapService;
    this.valueSetService = valueSetService;
    this.codeSystemService = codeSystemService;
  }

  /**
   * Execute a StructureMap on input content with optional validation
   */
  execute(structureMap: StructureMap, inputContent: any, options?: ExecutionOptions): EnhancedExecutionResult {
    try {
      // Basic validation
      if (!structureMap) {
        return {
          success: false,
          errors: ['StructureMap is required']
        };
      }

      if (!structureMap.group || structureMap.group.length === 0) {
        return {
          success: false,
          errors: ['StructureMap must have at least one group']
        };
      }

      const result: EnhancedExecutionResult = {
        success: true,
        result: undefined,
        validation: {}
      };

      // Validate input if requested
      if (options?.validateInput && options?.inputProfile) {
        const inputValidation = this.validationService.validate(inputContent, options.inputProfile);
        result.validation!.input = inputValidation;

        if (!inputValidation.valid && options?.strictMode) {
          return {
            success: false,
            errors: [`Input validation failed: ${inputValidation.errors.map(e => e.message).join(', ')}`],
            validation: result.validation
          };
        }
      }

      // Execute the main group
      const mainGroup = structureMap.group.find(g => g.name === 'main') || structureMap.group[0];
      const transformResult = this.executeGroup(mainGroup, inputContent);
      result.result = transformResult;

      // Validate output if requested
      if (options?.validateOutput && options?.outputProfile) {
        const outputValidation = this.validationService.validate(transformResult, options.outputProfile);
        result.validation!.output = outputValidation;

        if (!outputValidation.valid && options?.strictMode) {
          return {
            success: false,
            errors: [`Output validation failed: ${outputValidation.errors.map(e => e.message).join(', ')}`],
            validation: result.validation
          };
        }
      }

      return result;
    } catch (error) {
      return {
        success: false,
        errors: [error instanceof Error ? error.message : 'Unknown execution error']
      };
    }
  }

  /**
   * Get the validation service for registering StructureDefinitions
   */
  getValidationService(): ValidationService {
    return this.validationService;
  }

  /**
   * Execute a group within a StructureMap
   */
  private executeGroup(group: any, inputContent: any): any {
    // This is a basic implementation - a real StructureMap executor would be much more complex
    // and would need to handle FHIR Path expressions, complex transformations, etc.
    
    const result: any = {};

    // Process each rule in the group
    if (group.rule) {
      for (const rule of group.rule) {
        this.executeRule(rule, inputContent, result);
      }
    }

    return result;
  }

  /**
   * Execute a single mapping rule
   */
  private executeRule(rule: any, source: any, target: any): void {
    try {
      // Basic rule execution - map simple element to element
      if (rule.source && rule.target && rule.source.length > 0 && rule.target.length > 0) {
        const sourceElement = rule.source[0].element;
        const targetElement = rule.target[0].element;
        
        if (sourceElement && targetElement && source[sourceElement] !== undefined) {
          let value = source[sourceElement];
          
          // Check if target has transform operations
          const targetRule = rule.target[0];
          if (targetRule.transform) {
            value = this.applyTransform(targetRule.transform, value, targetRule.parameter);
          }
          
          target[targetElement] = value;
        }
      }
    } catch (error) {
      console.error('Error executing rule:', error);
    }
  }

  /**
   * Apply transform operations including terminology operations
   */
  private applyTransform(transform: string, value: any, parameters?: any[]): any {
    switch (transform) {
      case 'copy':
        return value;
        
      case 'translate':
        return this.applyTranslateTransform(value, parameters);
        
      case 'evaluate':
        // FHIRPath evaluation - basic implementation
        return this.evaluateFhirPath(value, parameters);
        
      case 'create':
        // Create a new resource/element
        return this.createResource(parameters);
        
      case 'reference':
        // Create a reference
        return this.createReference(value, parameters);
        
      case 'dateOp':
        // Date operations
        return this.applyDateOperation(value, parameters);
        
      case 'append':
        // String append operation
        return this.appendStrings(value, parameters);
        
      case 'cast':
        // Type casting
        return this.castValue(value, parameters);
        
      default:
        console.warn(`Unknown transform: ${transform}`);
        return value;
    }
  }

  /**
   * Apply translate transform using ConceptMaps
   */
  private applyTranslateTransform(value: any, parameters?: any[]): any {
    if (!this.conceptMapService || !parameters || parameters.length < 2) {
      return value;
    }

    try {
      const sourceSystem = parameters[0];
      const targetSystem = parameters[1];
      
      if (typeof value === 'object' && value.code && value.system) {
        // Handle Coding input
        const translations = this.conceptMapService.translate(
          value.system,
          value.code,
          targetSystem
        );
        
        if (translations.length > 0) {
          const translation = translations[0];
          return {
            system: translation.system || targetSystem,
            code: translation.code,
            display: translation.display
          };
        }
      } else if (typeof value === 'string') {
        // Handle string code input
        const translations = this.conceptMapService.translate(
          sourceSystem,
          value,
          targetSystem
        );
        
        if (translations.length > 0) {
          return translations[0].code;
        }
      }
    } catch (error) {
      console.error('Error in translate transform:', error);
    }
    
    return value;
  }

  /**
   * Basic FHIRPath evaluation
   */
  private evaluateFhirPath(value: any, parameters?: any[]): any {
    if (!parameters || parameters.length === 0) {
      return value;
    }
    
    const expression = parameters[0];
    
    // Very basic FHIRPath implementation - would need proper parser in production
    if (expression === 'true') return true;
    if (expression === 'false') return false;
    if (expression.startsWith("'") && expression.endsWith("'")) {
      return expression.slice(1, -1);
    }
    
    // Handle simple property access
    if (expression.includes('.')) {
      const parts = expression.split('.');
      let current = value;
      for (const part of parts) {
        if (current && typeof current === 'object') {
          current = current[part];
        } else {
          return undefined;
        }
      }
      return current;
    }
    
    return value;
  }

  /**
   * Create a new resource or element
   */
  private createResource(parameters?: any[]): any {
    if (!parameters || parameters.length === 0) {
      return {};
    }
    
    const resourceType = parameters[0];
    return { resourceType };
  }

  /**
   * Create a reference
   */
  private createReference(value: any, parameters?: any[]): any {
    if (typeof value === 'string') {
      return { reference: value };
    }
    
    if (value && value.resourceType && value.id) {
      return { reference: `${value.resourceType}/${value.id}` };
    }
    
    return value;
  }

  /**
   * Apply date operations
   */
  private applyDateOperation(value: any, parameters?: any[]): any {
    if (!parameters || parameters.length < 2) {
      return value;
    }
    
    const operation = parameters[0];
    const amount = parameters[1];
    
    try {
      const date = new Date(value);
      
      switch (operation) {
        case 'add':
          return new Date(date.getTime() + amount * 24 * 60 * 60 * 1000).toISOString();
        case 'subtract':
          return new Date(date.getTime() - amount * 24 * 60 * 60 * 1000).toISOString();
        case 'now':
          return new Date().toISOString();
        default:
          return value;
      }
    } catch (error) {
      return value;
    }
  }

  /**
   * Append strings
   */
  private appendStrings(value: any, parameters?: any[]): any {
    if (!parameters || parameters.length === 0) {
      return value;
    }
    
    let result = String(value || '');
    for (const param of parameters) {
      result += String(param);
    }
    
    return result;
  }

  /**
   * Cast value to different type
   */
  private castValue(value: any, parameters?: any[]): any {
    if (!parameters || parameters.length === 0) {
      return value;
    }
    
    const targetType = parameters[0];
    
    try {
      switch (targetType) {
        case 'string':
          return String(value);
        case 'integer':
          return parseInt(value, 10);
        case 'decimal':
          return parseFloat(value);
        case 'boolean':
          return Boolean(value);
        case 'date':
          return new Date(value).toISOString().split('T')[0];
        case 'dateTime':
          return new Date(value).toISOString();
        default:
          return value;
      }
    } catch (error) {
      return value;
    }
  }

  /**
   * Get terminology services for external access
   */
  getTerminologyServices(): {
    conceptMapService?: ConceptMapService;
    valueSetService?: ValueSetService;
    codeSystemService?: CodeSystemService;
  } {
    return {
      conceptMapService: this.conceptMapService,
      valueSetService: this.valueSetService,
      codeSystemService: this.codeSystemService
    };
  }

  /**
   * Validate that a StructureMap can be executed
   */
  validateStructureMap(structureMap: StructureMap): { valid: boolean; errors: string[] } {
    const errors: string[] = [];

    if (!structureMap) {
      errors.push('StructureMap is null or undefined');
      return { valid: false, errors };
    }

    if (structureMap.resourceType !== 'StructureMap') {
      errors.push('Resource type must be "StructureMap"');
    }

    if (!structureMap.group || structureMap.group.length === 0) {
      errors.push('StructureMap must have at least one group');
    }

    if (structureMap.group) {
      for (let i = 0; i < structureMap.group.length; i++) {
        const group = structureMap.group[i];
        if (!group.name) {
          errors.push(`Group ${i} must have a name`);
        }
        if (!group.input || group.input.length === 0) {
          errors.push(`Group ${i} must have at least one input`);
        }
      }
    }

    return { valid: errors.length === 0, errors };
  }
}

File: src/commonMain/kotlin/org/litlfred/fmlrunner/executor/StructureMapExecutor.kt
package org.litlfred.fmlrunner.executor

import org.litlfred.fmlrunner.types.*
import kotlinx.serialization.json.*
// kotlin-fhirpath for FHIRPath evaluation from https://github.com/jingtang10/kotlin-fhirpath
// Note: JitPack access still blocked - will integrate when network firewall allows
// import com.github.jingtang10.kotlin.fhirpath.FHIRPathEngine
// import com.github.jingtang10.kotlin.fhirpath.FHIRPathEngineFactory

/**
 * StructureMap execution engine - executes StructureMaps on input data
 * Ready for kotlin-fhirpath library integration when network access allows
 */
class StructureMapExecutor {
    // kotlin-fhirpath engine integration ready when network allows
    // private val fhirPathEngine: FHIRPathEngine = FHIRPathEngineFactory.create()
    
    /**
     * Execute a StructureMap on input content
     */
    fun execute(structureMap: StructureMap, inputContent: String, options: ExecutionOptions = ExecutionOptions()): ExecutionResult {
        return try {
            // Basic validation
            if (structureMap.group.isEmpty()) {
                return ExecutionResult(
                    success = false,
                    errors = listOf("StructureMap must have at least one group")
                )
            }

            // Parse input JSON
            val json = Json { ignoreUnknownKeys = true }
            val inputData = json.parseToJsonElement(inputContent)

            // Execute the main group (first group by convention)
            val mainGroup = structureMap.group.first()
            val result = executeGroup(mainGroup, inputData, options)

            ExecutionResult(
                success = true,
                result = Json.encodeToString(JsonElement.serializer(), result)
            )
        } catch (e: Exception) {
            ExecutionResult(
                success = false,
                errors = listOf("Execution failed: ${e.message}")
            )
        }
    }

    /**
     * Execute a StructureMap group
     */
    private fun executeGroup(group: StructureMapGroup, inputData: JsonElement, options: ExecutionOptions): JsonElement {
        // Create target data structure
        val targetData = JsonObject(emptyMap()).toMutableMap()

        // Find source and target inputs
        val sourceInput = group.input.find { it.mode == InputMode.SOURCE }
        val targetInput = group.input.find { it.mode == InputMode.TARGET }

        if (sourceInput == null || targetInput == null) {
            throw IllegalArgumentException("Group must have both source and target inputs")
        }

        // Create execution context
        val context = ExecutionContext(
            sourceData = inputData,
            targetData = targetData,
            sourceContext = sourceInput.name,
            targetContext = targetInput.name
        )

        // Execute rules
        for (rule in group.rule) {
            executeRule(rule, context, options)
        }

        return JsonObject(targetData)
    }

    /**
     * Execute a single rule
     */
    private fun executeRule(rule: StructureMapGroupRule, context: ExecutionContext, options: ExecutionOptions) {
        // Process each source
        for (source in rule.source) {
            val sourceValue = extractSourceValue(source, context)
            
            // Process targets if available
            rule.target?.forEach { target ->
                applyTargetMapping(source, target, sourceValue, context, options)
            }
        }
    }

    /**
     * Extract value from source
     */
    private fun extractSourceValue(source: StructureMapGroupRuleSource, context: ExecutionContext): JsonElement? {
        return try {
            when {
                source.context == context.sourceContext && source.element != null -> {
                    // Extract from source data
                    extractElementValue(context.sourceData, source.element)
                }
                source.context == context.sourceContext && source.element == null -> {
                    // Use entire source
                    context.sourceData
                }
                else -> null
            }
        } catch (e: Exception) {
            null
        }
    }

    /**
     * Apply target mapping
     */
    private fun applyTargetMapping(
        source: StructureMapGroupRuleSource,
        target: StructureMapGroupRuleTarget,
        sourceValue: JsonElement?,
        context: ExecutionContext,
        options: ExecutionOptions
    ) {
        if (sourceValue == null) return

        val targetElement = target.element ?: return
        
        // Apply transformation if specified
        val transformedValue = if (target.transform != null) {
            applyTransform(target.transform, sourceValue, target.parameter)
        } else {
            sourceValue
        }

        // Set target value
        setTargetValue(context, targetElement, transformedValue)
    }

    /**
     * Extract element value from JSON
     */
    private fun extractElementValue(jsonElement: JsonElement, elementPath: String): JsonElement? {
        return when (jsonElement) {
            is JsonObject -> jsonElement[elementPath]
            is JsonArray -> {
                // Handle array access
                val index = elementPath.toIntOrNull()
                if (index != null && index < jsonElement.size) {
                    jsonElement[index]
                } else null
            }
            else -> null
        }
    }

    /**
     * Set target value in context
     */
    private fun setTargetValue(context: ExecutionContext, elementPath: String, value: JsonElement) {
        context.targetData[elementPath] = value
    }

    /**
     * Apply transformation function
     */
    private fun applyTransform(transform: String, value: JsonElement, parameters: List<String>?): JsonElement {
        return when (transform.lowercase()) {
            "copy" -> value
            "create" -> {
                // Create new object/value based on parameters
                parameters?.firstOrNull()?.let { type ->
                    when (type.lowercase()) {
                        "string" -> JsonPrimitive("")
                        "integer" -> JsonPrimitive(0)
                        "boolean" -> JsonPrimitive(false)
                        else -> JsonObject(emptyMap())
                    }
                } ?: value
            }
            "cast" -> {
                // Type casting
                parameters?.firstOrNull()?.let { targetType ->
                    castValue(value, targetType)
                } ?: value
            }
            "evaluate" -> {
                // Simple FHIRPath-like evaluation (basic implementation)
                parameters?.firstOrNull()?.let { expression ->
                    evaluateExpression(value, expression)
                } ?: value
            }
            else -> value // Unknown transform, return original value
        }
    }

    /**
     * Cast value to target type
     */
    private fun castValue(value: JsonElement, targetType: String): JsonElement {
        return when (targetType.lowercase()) {
            "string" -> JsonPrimitive(value.toString().trim('"'))
            "integer" -> {
                val stringValue = if (value is JsonPrimitive) value.content else value.toString()
                JsonPrimitive(stringValue.toIntOrNull() ?: 0)
            }
            "boolean" -> {
                val stringValue = if (value is JsonPrimitive) value.content else value.toString()
                JsonPrimitive(stringValue.toBooleanStrictOrNull() ?: false)
            }
            else -> value
        }
    }

    /**
     * Expression evaluation prepared for kotlin-fhirpath integration
     */
    private fun evaluateExpression(context: JsonElement, expression: String): JsonElement {
        return try {
            // Prepared for kotlin-fhirpath engine when network access allows
            // val contextString = context.toString()
            // val results = fhirPathEngine.evaluate(contextString, expression)
            // if (results.isNotEmpty()) {
            //     Json.parseToJsonElement(results.first().toString())
            // } else {
            //     JsonNull
            // }
            
            // Enhanced fallback evaluation with improved FHIRPath-like support
            when {
                expression.startsWith("'") && expression.endsWith("'") -> {
                    // String literal
                    JsonPrimitive(expression.substring(1, expression.length - 1))
                }
                expression.matches(Regex("\\d+")) -> {
                    // Number literal
                    JsonPrimitive(expression.toInt())
                }
                expression == "true" || expression == "false" -> {
                    // Boolean literal
                    JsonPrimitive(expression.toBoolean())
                }
                expression.contains(".") -> {
                    // Property access path
                    val parts = expression.split(".")
                    var current = context
                    for (part in parts) {
                        current = extractElementValue(current, part) ?: return JsonNull
                    }
                    current
                }
                else -> {
                    // Simple property access
                    extractElementValue(context, expression) ?: JsonNull
                }
            }
        } catch (e: Exception) {
            JsonNull
        }
    }

    /**
     * Validate StructureMap structure
     */
    fun validateStructureMap(structureMap: StructureMap): ValidationResult {
        val errors = mutableListOf<String>()

        if (structureMap.group.isEmpty()) {
            errors.add("StructureMap must have at least one group")
        }

        for ((index, group) in structureMap.group.withIndex()) {
            if (group.name.isBlank()) {
                errors.add("Group $index must have a name")
            }

            if (group.input.isEmpty()) {
                errors.add("Group '${group.name}' must have at least one input")
            }

            val sourceInputs = group.input.filter { it.mode == InputMode.SOURCE }
            val targetInputs = group.input.filter { it.mode == InputMode.TARGET }

            if (sourceInputs.isEmpty()) {
                errors.add("Group '${group.name}' must have at least one source input")
            }

            if (targetInputs.isEmpty()) {
                errors.add("Group '${group.name}' must have at least one target input")
            }

            if (group.rule.isEmpty()) {
                errors.add("Group '${group.name}' must have at least one rule")
            }

            for ((ruleIndex, rule) in group.rule.withIndex()) {
                if (rule.source.isEmpty()) {
                    errors.add("Rule $ruleIndex in group '${group.name}' must have at least one source")
                }
            }
        }

        return ValidationResult(
            valid = errors.isEmpty(),
            errors = errors
        )
    }
}

/**
 * Execution context for rule processing
 */
private data class ExecutionContext(
    val sourceData: JsonElement,
    val targetData: MutableMap<String, JsonElement>,
    val sourceContext: String,
    val targetContext: String
)

/**
 * Validation result
 */
data class ValidationResult(
    val valid: Boolean,
    val errors: List<String>
)